pkg/model/diamond.go

```diff
@@ -0,0 +1,215 @@
+	DiamondCancelled DiamondState = "cancelled"
+)
+
+const (
+	diamondDescriptorFile = "diamond.yaml"
+	splitDescriptorFile   = "split.yaml"
+)
+
+// IsValid checks the value of a diamond state
+func (s DiamondState) IsValid() bool {
+	switch s {
+	case DiamondInitialized, DiamondRunning, DiamondCommitting, DiamondDone, DiamondCancelling, DiamondCancelled:
+		return true
+	default:
+		return false
+	}
+}
+
+func (s DiamondState) String() string {
+	return string(s)
+}
+
+// SplitState models the running status of an ungoing diamond split
+type SplitState string
+
+const (
+	// SplitDone is the state of a completed split
+	SplitDone SplitState = "done"
+
+	// SplitFailed is the state of a failed split
+	SplitFailed SplitState = "failed"
+
+	// SplitRunning is the state of running split
+	SplitRunning SplitState = "running"
+)
+
+// IsValid checks the value of a diamond state
+func (s SplitState) IsValid() bool {
+	switch s {
+	case SplitDone, SplitFailed, SplitRunning:
+		return true
+	default:
+		return false
+	}
+}
+
+func (s SplitState) String() string {
+	return string(s)
+}
+
+// DiamondDescriptor models a diamond's metadata
+type DiamondDescriptor struct {
+	DiamondID       string            `json:"diamondID" yaml:"splitID"`
+	StartTime       time.Time         `json:"startTime" yaml:"starTime"` // when the diamond was initialized
+	EndTime         time.Time         `json:"endTime,omitempty" yaml:"endTime,omitempty"`
+	StateChangeTime time.Time         `json:"stateChangeTime" yaml:"state"` // last state change
+	State           DiamondState      `json:"state" yaml:"state"`
+	HasConflicts    bool              `json:"hasConflicts,omitempty" yaml:"hasConflicts,omitempty"`
+	HasCheckpoints  bool              `json:"hasCheckpoints,omitempty" yaml:"hasCheckpoints,omitempty"`
+	Tag             string            `json:"tag,omitempty" yaml:"tag,omitempty"`
+	Splits          []SplitDescriptor `json:"splits,omitempty" yaml:"splits,omitempty"`
+	_               struct{}
+}
+
+func defaultDiamondDescriptor() *DiamondDescriptor {
+	ksuid, err := ksuid.NewRandom()
+	if err != nil {
+		panic(fmt.Sprintf("cannot generate random ksuid: %v", err))
+	}
+	return &DiamondDescriptor{
+		DiamondID: ksuid.String(),
+		StartTime: GetBundleTimeStamp(),
+		State:     DiamondInitialized,
+	}
+}
+
+// NewDiamondDescriptor builds a new DiamondDescriptor
+func NewDiamondDescriptor(opts ...DiamondDescriptorOption) *DiamondDescriptor {
+	d := defaultDiamondDescriptor()
+	for _, apply := range opts {
+		apply(d)
+	}
+	return d
+}
+
+// SplitDescriptor models the metadata about a given split within a diamond
+type SplitDescriptor struct {
+	SplitID      string        `json:"splitID" yaml:"splitID"`
+	DiamondID    string        `json:"diamondID" yaml:"splitID"`
+	StartTime    time.Time     `json:"startTime" yaml:"starTime"`
+	EndTime      time.Time     `json:"endTime,omitempty" yaml:"endTime,omitempty"`
+	State        SplitState    `json:"state" yaml:"state"`
```

Comment:
swooping in kind of early with comments here in order to understand design proposal as much as talk about implementation details .. was i was trying to ask in my comments about `split add` is why `State` is appearing as a struct field rather than a method.

related but not equivalent question is why there are distinct `running` and `failed` states.  and this definitely veers to implementation detail:  don't even understand how we could distinguish in-progress `add`s from `failed` adds in all cases, such as if the internet connection cuts out during the 2KiB slices' upload. 

i suspect this'll be clarified by actual implementation and is best discussed after more source is written.

Reply:
> why State is appearing as a struct field rather than a method.

state is persisted

>  why there are distinct running and failed states. and this definitely veers to implementation detail: don't even understand how we could distinguish in-progress adds from failed adds in all cases, 

Design

    Diamonds abide by the following state-transition model:
        initialized
        committing
        done|cancelled

    Splits:
        running
        done|failed

    NOTE: the cancelled and failed states are not critical to the functionality: they merely provide better monitoring capabilities (""what's going on with my splits?") as well as safe cleaning up.


### Failed add operation

A controlled failure on such an operation would mark the split as explicitly failed (for further reporting and monitoring).

/splits/{repo}/{diamond-id}/{split-id}/split.yaml   #  <- State pass to "failed"

Please mark that it is not possible to distinguish an aborted or hanging add operation from a merely long running one.

Reply:
since there's a "controlled failure", there's also an "uncontrolled failure" that also can be distinguished from aborted, hanging, or long-running adds?

Reply:
yes indeed

Reply:
ok, so i guess this was my fundamental qualm with having the failure statuses:

how much of the space of all possible failures (both controlled and uncontrolled) needs to be covered by controlled failures in order for controlled failures to be meaningful?

by now, the decision-making is veering toward the philosophical, so it's probably time to let the overall design play out in source while keeping the more abstruse considerations in mind for future revisions?

marking this as resolved in any case.